Objective-C/CBLCollection.mm

```diff
@@ -21,45 +21,139 @@
 #import "CBLIndexable.h"
 #import "CBLChangeListenerToken.h"
 #import "CBLCollectionChangeObservable.h"
+#import "CBLDatabase+Internal.h"
+#import "CBLStringBytes.h"
+#import "CBLScope.h"
+#import "CBLIndexConfiguration+Internal.h"
+#import "CBLStatus.h"
+#import "CBLScope+Internal.h"
+
+using namespace fleece;
 
 NSString* const kCBLDefaultCollectionName = @"_default";
 
-@implementation CBLCollection
+@implementation CBLCollection {
+    __weak CBLDatabase* _db;
+    C4Collection* _c4col;
+}
 
 @synthesize count=_count, name=_name, scope=_scope;
 
-- (instancetype) initWithName: (NSString*)name
-                        scope: (nullable CBLScope*)scope
-                        error: (NSError**)error {
-    CBLAssertNotNil(name);
+- (instancetype) initWithDB: (CBLDatabase*)db
+             collectionName: (NSString*)collectionName
+                  scopeName: (NSString*)scopeName
+                      error: (NSError**)error {
+    CBLAssertNotNil(db);
+    CBLAssertNotNil(collectionName);
     
     self = [super init];
     if (self) {
-        _name = name;
-        if (scope)
-            _scope = scope;
+        _db = db;
+        _name = collectionName;
+        
+        scopeName = scopeName.length > 0 ? scopeName : kCBLDefaultScopeName;
+        CBLStringBytes cName(collectionName);
+        CBLStringBytes sName(scopeName);
+        C4CollectionSpec spec = { .name = cName, .scope = sName };
+        
+        __block C4Collection* c;
+        __block C4Error err = {};
+        [db safeBlock:^{
+            c = c4db_getCollection(db.c4db, spec);
+            if (!c) {
+                // collection doesn't exists, create one
+                c = c4db_createCollection(db.c4db, spec, &err);
+            }
+        }];
+        
+        if (err.code != 0) {
+            // TODO: Add to CBLErrorMessage : CBL-3296
+            CBLWarn(Database, @"%@ Failed to create new collection (%d/%d)",
+                    self, err.code, err.domain);
+            convertError(err, error);
+            return nil;
+        }
+        
+        _scope = [[CBLScope alloc] initWithDB: db name: scopeName error: error];
+        
+        _c4col = c;
     }
     return self;
 }
 
+- (void) dealloc {
+    _db = nil;
+    _c4col = nil;
+}
+
+- (NSString*) description {
+    return [NSString stringWithFormat: @"%@[%@.%@]%@", self.class, _scope.name, _name, _db];
+}
+
+#pragma mark - Indexable
+
 - (BOOL) createIndexWithName: (NSString*)name
                       config: (CBLIndexConfiguration *)config
                        error: (NSError**)error {
-    // TODO: add implementation
-    return NO;
+    CBLAssertNotNil(name);
+    CBLAssertNotNil(config);
+    
+    CBLDatabase* db = _db;
+    __block BOOL success = NO;
+    __block C4Error c4err;
+    [db safeBlock: ^{
```

Comment:
You would need to take care two things:
1. When the db pointer is not valid anymore.
2. When database is closed. So check Open is still needed.

Reply:
Probably checking if the collection is valid might be enough? I think there is a PR in LiteCore that tries to handle this but not at the current master.

Reply:
collectionIsValid check done!